x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/RollupDataExtractorFactory.java

```diff
@@ -0,0 +1,219 @@
+
+        final long datafeedInterval = getHistogramIntervalMillis(datafeedHistogramAggregation);
+
+        List<ParsedRollupCaps> validIntervalCaps = rollupCapsSet.stream()
+            .filter(rollupCaps -> validInterval(datafeedInterval, rollupCaps))
+            .collect(Collectors.toList());
+
+        if (validIntervalCaps.isEmpty()) {
+            listener.onFailure(
+                new IllegalArgumentException(
+                    "Rollup capabilities do not have a [date_histogram] aggregation with an interval " +
+                        "that is a multiple of the datafeed's interval.")
+            );
+            return;
+        }
+        final List<AggregationBuilder> flattenedAggs = new ArrayList<>();
+        flattenAggregations(datafeed.getAggregations().getAggregatorFactories(), datafeedHistogramAggregation, flattenedAggs);
+
+        if (validIntervalCaps.stream().noneMatch(rollupJobConfig -> hasAggregations(rollupJobConfig, flattenedAggs))) {
+            listener.onFailure(
+                new IllegalArgumentException("Rollup capabilities do not support all the datafeed aggregations at the desired interval.")
+            );
+            return;
+        }
+
+        listener.onResponse(new RollupDataExtractorFactory(client, datafeed, job));
+    }
+
+    private static boolean validInterval(long datafeedInterval, ParsedRollupCaps rollupJobGroupConfig) {
+        if (rollupJobGroupConfig.hasDatehistogram() == false) {
+            return false;
+        }
+        if (DateTimeZone.UTC.toString().equalsIgnoreCase(rollupJobGroupConfig.getTimezone()) == false) {
+            return false;
+        }
+        try {
+            long jobInterval = validateAndGetCalendarInterval(rollupJobGroupConfig.getInterval());
+            return datafeedInterval % jobInterval == 0;
+        } catch (ElasticsearchStatusException exception) {
+            return false;
+        }
+    }
+
+    private static void flattenAggregations(final Collection<AggregationBuilder> datafeedAggregations,
+                                            final AggregationBuilder datafeedHistogramAggregation,
+                                            final List<AggregationBuilder> flattenedAggregations) {
+        for (AggregationBuilder aggregationBuilder : datafeedAggregations) {
+            if (aggregationBuilder.equals(datafeedHistogramAggregation) == false) {
+                flattenedAggregations.add(aggregationBuilder);
+            }
+            flattenAggregations(aggregationBuilder.getSubAggregations(), datafeedHistogramAggregation, flattenedAggregations);
+        }
+    }
+
+    private static boolean hasAggregations(ParsedRollupCaps rollupCaps, List<AggregationBuilder> datafeedAggregations) {
+        for (AggregationBuilder aggregationBuilder : datafeedAggregations) {
+            String type = aggregationBuilder.getType();
+            String field = ((ValuesSourceAggregationBuilder) aggregationBuilder).field();
```

Comment:
I think it would be cleaner if the parameter of this method is already `List<ValuesSourceAggregationBuilder>` and if the calling code would therefore downcast as soon as it verified/knows that the `AggregationBuilder` is a `ValuesSourceAggregationBuilder` (I assume anything else would be a bug anyway).